File: trinity-nexus/lang/libs/kotlin/trinity-vsa/src/main/kotlin/com/trinity/vsa/TrinityVSA.kt
package com.trinity.vsa

import kotlin.math.sqrt
import kotlin.random.Random

/** Balanced ternary value: -1, 0, or +1 */
typealias Trit = Byte

const val TRIT_NEG: Trit = -1
const val TRIT_ZERO: Trit = 0
const val TRIT_POS: Trit = 1

/** Dense vector of balanced ternary values */
class TritVector(val data: ByteArray) {
    val dim: Int get() = data.size

    operator fun get(i: Int): Trit = data[i]
    operator fun set(i: Int, value: Trit) { data[i] = value }

    /** Number of non-zero elements */
    fun nnz(): Int = data.count { it != TRIT_ZERO }

    /** Sparsity (fraction of zeros) */
    fun sparsity(): Double = 1.0 - nnz().toDouble() / dim

    /** Negate in place */
    fun negate() {
        for (i in data.indices) data[i] = (-data[i]).toByte()
    }

    /** Clone vector */
    fun copy(): TritVector = TritVector(data.copyOf())

    companion object {
        /** Create zero vector */
        fun zeros(dim: Int) = TritVector(ByteArray(dim))

        /** Create random vector with balanced distribution */
        fun random(dim: Int, seed: Long? = null): TritVector {
            val rng = seed?.let { Random(it) } ?: Random
            val data = ByteArray(dim) { (rng.nextInt(3) - 1).toByte() }
            return TritVector(data)
        }
    }
}

/** Bind two vectors (element-wise multiplication) */
fun bind(a: TritVector, b: TritVector): TritVector {
    require(a.dim == b.dim) { "Dimension mismatch" }
    val result = ByteArray(a.dim) { i -> (a.data[i] * b.data[i]).toByte() }
    return TritVector(result)
}

/** Unbind (inverse of bind) */
fun unbind(a: TritVector, b: TritVector) = bind(a, b)

/** Bundle multiple vectors via majority voting */
fun bundle(vectors: List<TritVector>): TritVector {
    require(vectors.isNotEmpty()) { "Empty vector list" }
    val dim = vectors[0].dim
    val result = ByteArray(dim)
    for (i in 0 until dim) {
        val sum = vectors.sumOf { it.data[i].toInt() }
        result[i] = when {
            sum > 0 -> TRIT_POS
            sum < 0 -> TRIT_NEG
            else -> TRIT_ZERO
        }
    }
    return TritVector(result)
}

/** Circular permutation */
fun permute(v: TritVector, shift: Int): TritVector {
    val result = ByteArray(v.dim)
    for (i in 0 until v.dim) {
        val newIdx = Math.floorMod(i + shift, v.dim)
        result[newIdx] = v.data[i]
    }
    return TritVector(result)
}

/** Dot product */
fun dot(a: TritVector, b: TritVector): Long {
    require(a.dim == b.dim) { "Dimension mismatch" }
    return a.data.indices.sumOf { a.data[it].toLong() * b.data[it].toLong() }
}

/** Cosine similarity */
fun similarity(a: TritVector, b: TritVector): Double {
    val d = dot(a, b).toDouble()
    val normA = sqrt(dot(a, a).toDouble())
    val normB = sqrt(dot(b, b).toDouble())
    return if (normA == 0.0 || normB == 0.0) 0.0 else d / (normA * normB)
}

/** Hamming distance */
fun hammingDistance(a: TritVector, b: TritVector): Int {
    require(a.dim == b.dim) { "Dimension mismatch" }
    return a.data.indices.count { a.data[it] != b.data[it] }
}

/** Packed trit vector using bitsliced storage */
class PackedTritVec(val pos: LongArray, val neg: LongArray, val dim: Int) {
    val numWords: Int get() = pos.size

    companion object {
        fun from(v: TritVector): PackedTritVec {
            val numWords = (v.dim + 63) / 64
            val pos = LongArray(numWords)
            val neg = LongArray(numWords)
            for (i in 0 until v.dim) {
                val wordIdx = i / 64
                val bitIdx = i % 64
                when (v.data[i]) {
                    TRIT_POS -> pos[wordIdx] = pos[wordIdx] or (1L shl bitIdx)
                    TRIT_NEG -> neg[wordIdx] = neg[wordIdx] or (1L shl bitIdx)
                }
            }
            return PackedTritVec(pos, neg, v.dim)
        }
    }

    fun toVector(): TritVector {
        val data = ByteArray(dim)
        for (i in 0 until dim) {
            val wordIdx = i / 64
            val bitIdx = i % 64
            val posSet = (pos[wordIdx] shr bitIdx) and 1L
            val negSet = (neg[wordIdx] shr bitIdx) and 1L
            data[i] = when {
                posSet == 1L -> TRIT_POS
                negSet == 1L -> TRIT_NEG
                else -> TRIT_ZERO
            }
        }
        return TritVector(data)
    }
}

/** Fast packed bind */
fun packedBind(a: PackedTritVec, b: PackedTritVec): PackedTritVec {
    require(a.dim == b.dim) { "Dimension mismatch" }
    val pos = LongArray(a.numWords)
    val neg = LongArray(a.numWords)
    for (i in 0 until a.numWords) {
        pos[i] = (a.pos[i] and b.pos[i]) or (a.neg[i] and b.neg[i])
        neg[i] = (a.pos[i] and b.neg[i]) or (a.neg[i] and b.pos[i])
    }
    return PackedTritVec(pos, neg, a.dim)
}

/** Fast packed dot product */
fun packedDot(a: PackedTritVec, b: PackedTritVec): Long {
    require(a.dim == b.dim) { "Dimension mismatch" }
    var posCount = 0L
    var negCount = 0L
    for (i in 0 until a.numWords) {
        posCount += java.lang.Long.bitCount((a.pos[i] and b.pos[i]) or (a.neg[i] and b.neg[i]))
        negCount += java.lang.Long.bitCount((a.pos[i] and b.neg[i]) or (a.neg[i] and b.pos[i]))
    }
    return posCount - negCount
}

File: src/libs/scala/trinity-vsa/src/main/scala/com/trinity/vsa/TrinityVSA.scala
package com.trinity.vsa

import scala.util.Random
import scala.math.sqrt

/** Balanced ternary value: -1, 0, or +1 */
type Trit = Byte

object Trit:
  val Neg: Trit = -1
  val Zero: Trit = 0
  val Pos: Trit = 1

/** Dense vector of balanced ternary values */
class TritVector(val data: Array[Byte]):
  def dim: Int = data.length
  def apply(i: Int): Trit = data(i)
  def update(i: Int, v: Trit): Unit = data(i) = v
  
  def nnz: Int = data.count(_ != 0)
  def sparsity: Double = 1.0 - nnz.toDouble / dim
  
  def negate(): Unit = 
    for i <- data.indices do data(i) = (-data(i)).toByte
  
  def copy: TritVector = TritVector(data.clone())

object TritVector:
  def zeros(dim: Int): TritVector = TritVector(new Array[Byte](dim))
  
  def random(dim: Int, seed: Option[Long] = None): TritVector =
    val rng = seed.map(new Random(_)).getOrElse(new Random())
    val data = Array.fill(dim)((rng.nextInt(3) - 1).toByte)
    TritVector(data)

object VSA:
  /** Bind two vectors */
  def bind(a: TritVector, b: TritVector): TritVector =
    require(a.dim == b.dim, "Dimension mismatch")
    TritVector(Array.tabulate(a.dim)(i => (a.data(i) * b.data(i)).toByte))
  
  /** Unbind (inverse of bind) */
  def unbind(a: TritVector, b: TritVector): TritVector = bind(a, b)
  
  /** Bundle via majority voting */
  def bundle(vectors: Seq[TritVector]): TritVector =
    require(vectors.nonEmpty, "Empty vector list")
    val dim = vectors.head.dim
    val result = new Array[Byte](dim)
    for i <- 0 until dim do
      val sum = vectors.map(_.data(i).toInt).sum
      result(i) = if sum > 0 then Trit.Pos else if sum < 0 then Trit.Neg else Trit.Zero
    TritVector(result)
  
  /** Circular permutation */
  def permute(v: TritVector, shift: Int): TritVector =
    val result = new Array[Byte](v.dim)
    for i <- 0 until v.dim do
      val newIdx = Math.floorMod(i + shift, v.dim)
      result(newIdx) = v.data(i)
    TritVector(result)
  
  /** Dot product */
  def dot(a: TritVector, b: TritVector): Long =
    require(a.dim == b.dim, "Dimension mismatch")
    (0 until a.dim).map(i => a.data(i).toLong * b.data(i).toLong).sum
  
  /** Cosine similarity */
  def similarity(a: TritVector, b: TritVector): Double =
    val d = dot(a, b).toDouble
    val normA = sqrt(dot(a, a).toDouble)
    val normB = sqrt(dot(b, b).toDouble)
    if normA == 0 || normB == 0 then 0.0 else d / (normA * normB)
  
  /** Hamming distance */
  def hammingDistance(a: TritVector, b: TritVector): Int =
    require(a.dim == b.dim, "Dimension mismatch")
    (0 until a.dim).count(i => a.data(i) != b.data(i))

/** Packed trit vector */
class PackedTritVec(val pos: Array[Long], val neg: Array[Long], val dim: Int):
  def numWords: Int = pos.length
  
  def toVector: TritVector =
    val data = new Array[Byte](dim)
    for i <- 0 until dim do
      val wordIdx = i / 64
      val bitIdx = i % 64
      val posSet = (pos(wordIdx) >> bitIdx) & 1L
      val negSet = (neg(wordIdx) >> bitIdx) & 1L
      data(i) = if posSet == 1 then Trit.Pos else if negSet == 1 then Trit.Neg else Trit.Zero
    TritVector(data)

object PackedTritVec:
  def from(v: TritVector): PackedTritVec =
    val numWords = (v.dim + 63) / 64
    val pos = new Array[Long](numWords)
    val neg = new Array[Long](numWords)
    for i <- 0 until v.dim do
      val wordIdx = i / 64
      val bitIdx = i % 64
      v.data(i) match
        case Trit.Pos => pos(wordIdx) |= 1L << bitIdx
        case Trit.Neg => neg(wordIdx) |= 1L << bitIdx
        case _ =>
    PackedTritVec(pos, neg, v.dim)
  
  def bind(a: PackedTritVec, b: PackedTritVec): PackedTritVec =
    require(a.dim == b.dim, "Dimension mismatch")
    val pos = Array.tabulate(a.numWords)(i => (a.pos(i) & b.pos(i)) | (a.neg(i) & b.neg(i)))
    val neg = Array.tabulate(a.numWords)(i => (a.pos(i) & b.neg(i)) | (a.neg(i) & b.pos(i)))
    PackedTritVec(pos, neg, a.dim)
  
  def dot(a: PackedTritVec, b: PackedTritVec): Long =
    require(a.dim == b.dim, "Dimension mismatch")
    var posCount = 0L
    var negCount = 0L
    for i <- 0 until a.numWords do
      posCount += java.lang.Long.bitCount((a.pos(i) & b.pos(i)) | (a.neg(i) & b.neg(i)))
      negCount += java.lang.Long.bitCount((a.pos(i) & b.neg(i)) | (a.neg(i) & b.pos(i)))
    posCount - negCount
